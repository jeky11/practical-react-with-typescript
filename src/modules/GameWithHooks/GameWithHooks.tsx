import React, { FC, useCallback } from 'react';
import { useSearchParams } from 'react-router-dom';

import { GameLevels, LevelNames } from '@/modules/GameSettings';

import { Scoreboard } from '@/components/scoreboard';
import { Grid } from '@/components/grid';
import { GameOver } from '@/components/game';

import { useGame } from './useGame';

export const GameWithHooks: FC = () => {
  const [searchParams, setSearchParams] = useSearchParams();

  const urlLevelParam = (searchParams.get('level') || undefined) as LevelNames;

  const {
    level,
    time,
    isGameOver,
    isWin,
    settings,
    playerField,
    flagCounter,
    onClick,
    onContextMenu,
    onChangeLevel,
    onReset,
  } = useGame(urlLevelParam);

  const [, bombs] = settings;

  const onChangeLevelHandler = useCallback(
    ({ target: { value: level } }: React.ChangeEvent<HTMLSelectElement>) => {
      setSearchParams({ level });
      onChangeLevel(level as LevelNames);
    },
    // Stryker disable next-line ArrayDeclaration
    []
  );

  return (
    <>
      <Scoreboard
        time={String(time)}
        bombs={String(bombs - flagCounter)}
        levels={GameLevels as unknown as string[]}
        defaultLevel={level}
        onChangeLevel={onChangeLevelHandler}
        onReset={onReset}
      />
      {isGameOver && <GameOver onClick={onReset} isWin={isWin} />}
      <Grid onClick={onClick} onContextMenu={onContextMenu}>
        {playerField}
      </Grid>
    </>
  );
};
