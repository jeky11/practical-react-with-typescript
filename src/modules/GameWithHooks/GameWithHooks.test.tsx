import React from 'react';
import { useSearchParams } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';

import { GameWithHooks } from './GameWithHooks';

const mockOnClick = jest.fn();
const mockOnChangeLevel = jest.fn();
const mockOnReset = jest.fn();
const mockOnContextMenu = jest.fn();
const mockSetSearchParams = jest.fn();

jest.mock('react-router-dom', () => ({
  useSearchParams: jest.fn(),
}));

(useSearchParams as jest.Mock).mockReturnValue([
  { get: () => null },
  mockSetSearchParams,
]);

jest.mock('./useGame', () => ({
  __esModule: true,
  useGame: (level = 'beginner') => ({
    level,
    time: 0,
    flagCounter: 0,
    isGameOver: true,
    isWin: false,
    settings: [9, 10],
    playerField: [
      [10, 10],
      [10, 10],
    ],
    onClick: mockOnClick,
    onContextMenu: mockOnContextMenu,
    onChangeLevel: mockOnChangeLevel,
    onReset: mockOnReset,
  }),
}));

beforeEach(() => {
  jest.clearAllMocks();
});

describe('GameWithHooks test cases', () => {
  it('Render game field by default', () => {
    const { asFragment } = render(<GameWithHooks />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('Cell click works fine', () => {
    render(<GameWithHooks />);
    userEvent.click(screen.getByTestId('0,0'));
    expect(mockOnClick).toHaveBeenCalled();
  });
  it('Context menu handler on a cell works fine', () => {
    render(<GameWithHooks />);
    userEvent.click(screen.getByTestId('0,0'), { button: 2 });
    expect(mockOnContextMenu).toHaveBeenCalled();
  });
  it('Reset handler works fine', () => {
    render(<GameWithHooks />);
    userEvent.click(screen.getByRole('button'));
    expect(mockOnReset).toHaveBeenCalled();
  });
  it('Change level works fine', () => {
    render(<GameWithHooks />);
    userEvent.selectOptions(screen.getByRole('combobox'), 'intermediate');
    expect(mockOnChangeLevel).toHaveBeenCalled();
    expect(mockSetSearchParams).toHaveBeenCalledWith({ level: 'intermediate' });
  });
  it('Level in search params works fine', () => {
    (useSearchParams as jest.Mock).mockReturnValue([
      { get: () => 'intermediate' },
    ]);
    render(<GameWithHooks />);
    const intermediateOption = screen.queryByText('intermediate');
    expect(intermediateOption).toBeInTheDocument();
  });
  it('Game over reset the game state', () => {
    render(<GameWithHooks />);
    userEvent.click(screen.getByText('🙁'));
    expect(mockOnReset).toHaveBeenCalled();
  });
});
