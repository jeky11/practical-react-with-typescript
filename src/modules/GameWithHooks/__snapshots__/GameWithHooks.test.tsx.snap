// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameWithHooks test cases Render game field by default 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

.emotion-1 {
  display: inline-block;
  padding: 0 0.3vw;
  color: #ec433c;
  border: 0.15vw inset;
  line-height: 2vw;
  letter-spacing: 0.08em;
  background: #333;
  text-shadow: 0 0 0.1vw #ec433c;
}

.emotion-2 {
  margin: 0;
  height: 100%;
  border-radius: 0;
  border: 0.15vw solid;
  border-color: white #9e9e9e #9e9e9e white;
  background-color: #d1d1d1;
}

.emotion-3 {
  font-weight: normal;
  display: block;
  white-space: nowrap;
  min-height: 1.2em;
  padding: 0 0.2vw 0.2vw;
}

.emotion-6 {
  font-size: 1.1vw;
  height: 100%;
  cursor: pointer;
  font-weight: 700;
  border-width: 0.15vw;
  border-style: solid;
  background-color: #d1d1d1;
  border-color: white #9e9e9e #9e9e9e white;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      0
    </div>
    <div>
      <select
        class="emotion-2"
      >
        <option
          class="emotion-3"
          value="beginner"
        >
          beginner
        </option>
        <option
          class="emotion-3"
          value="intermediate"
        >
          intermediate
        </option>
        <option
          class="emotion-3"
          value="expert"
        >
          expert
        </option>
      </select>
      <button
        class="emotion-6"
      >
        üôÇ
      </button>
    </div>
    <div
      class="emotion-1"
    >
      10
    </div>
  </div>
  .emotion-0 {
  top: 60%;
  left: 50%;
  z-index: 11;
  width: 8vw;
  height: 8vw;
  font-size: 4vw;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  line-height: 8vw;
  position: absolute;
  border-radius: 50%;
  pointer-events: auto;
  background-color: #d1d1d1;
  -webkit-transform: translate(-50%, -50%);
  -moz-transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  border: 1px solid rgba(51, 51, 51, 0.25);
}

<div
    class="emotion-0"
  >
    üôÅ
  </div>
  .emotion-0 {
  display: grid;
  grid-template-columns: repeat(2, auto);
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  padding: 1vw;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  width: 1.8vh;
  height: 1.8vh;
  color: transparent;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: white #9e9e9e #9e9e9e white;
}

.emotion-1:hover {
  -webkit-filter: brightness(1.1);
  filter: brightness(1.1);
}

<div
    class="emotion-0"
    role="grid"
    size="2"
  >
    <div
      class="emotion-1"
      data-testid="0,0"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="0,1"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="1,0"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="1,1"
      role="cell"
    >
      10
    </div>
  </div>
</DocumentFragment>
`;
